shared/src/bad_token/trace_call.rs

```diff
@@ -182,20 +182,22 @@ impl TraceCallDetector {
             .public_address()
     }
 
-    fn create_trace_request(&self, token: H160, amount: U256, take_from: H160) -> Vec<CallRequest> {
+    fn create_trace_request(&self, token: H160, amount: U256, pool: H160) -> Vec<CallRequest> {
```

Comment:
The from address here is not necessarily a uniswap pool. What we want is *any* address that owns the token. This used to be only our uniswap like liquidity sources but also sushiswap and has more recently also become the balancer vault contract because it holds many different tokens.
I don't mind adding an explicit check for transferring to a uniswap pool but we should make it more explicit than this. Could be another function `check_transfer_to_uniswap`. Also since there can be multiple pools that contain the token is there a way to pick a specific one?

Reply:
`take_from` comes from `find_largest_pool_owning_token`, so I imagine it's always a pool of some sort. To me checking if you can trade it in the largest pool makes sense, even if the exact pool used might change in the future.
Is the problem the fact that other pools might not work by using a direct transfer, and so transferring might fail even if this is not an issue with the current pool? I could maybe add a filter to `find_largest_pool_owning_token` that only makes it return pools that use a single transfer and give to `create_trace_request` both `take_from` and `pool`.

Reply:
So what I would do is adding a new function `type` to the trait `AmmPairProvider` that returns `enum Type {Uniswap, Sushiswap, /* evenutally Balancer */ }` and change `find_largest_pool_owning_token` so that gives you a pool of `Type::Uniswap` or `Type::Sushiswap` if possible, and if not the transfer to pool check is skipped. Does it sound reasonable?

Reply:
The name find_largest_pool_owning_token is outdated because it can also return the balancer vault (afaik this is shared between many pools).
What I am getting at is that this was originally designed just to find **any** address that owns the token and using the pools was a natural way to do that. But now we are saying this is explicitly supposed be the largest uniswap (?) pool.
I think it would be cleaner to separate these two checks a bit. For example by passing in an optional largest pool that should also be transferred to (which can be same as take_from but doesn't have to be) or doing it in another function that does its own trace_call. Then also have a separate function for explicitly finding the largest pool which can share implementation details with the existing function that should really be called find_any_address_that_owns_token.
I'm not sure about the best design from the top of my head. I feel like we don't have to change AmmPairProvider because it is supposed to just return uniswap-like pools and for this case the largest uniswap-like pool should be fine? It becomes more hairy when we consider balancer, uniswapv2 vs v3 and whether we should just check the largest pool or all of them.

Btw where in the code does that token prevent the transfer? I couldn't find it.

Reply:
Makes sense, I'll try to rework this PR with these comments in mind.
About the concept of "largest": I'm personally ok with just finding an educated guess for "largest" and have it restricted to Uniswap/Sushiswap. My idea is that a token that implements weird restrictions on some pools is probably not really going to be successful and have many pools, so that "largest"  can be fairly well defined by a guesswork heuristic (maybe "only" would be enough). If not, the token is going to need special attention, but it's not worse than before this PR. Of the 94 bad tokens in the test, 9 fail the new check, 3 of which did non fail any other check before, which I consider a good enough result.
The transfer is prevented [here](https://etherscan.io/address/0xcfbD04B3ceF2CF1527F143A49e5dc1e19941d254#code#L530), with the additional knowledge that `burnAddress` is the Uniswap pair WETH/token! 

What I'd do now:
1. `find_largest_pool_owning_token` becomes a standalone function `find_largest_address_owning_token`.
2. two new functions for `TraceCallDetector`: `find_rich_token_holder` (returning the same as the current `find_largest_pool_owning_token`) and `find_large_pool_owning_token` (returning the same as the current `find_largest_pool_owning_token` with the restriction that the address comes from only Uniswap and Sushiswap pools).
3. If there is no Sushiswap/Uniswap pool, this particular test is skipped.
3. Ideally I make this efficient so that no balance is queried twice, so it probably won't look exactly as I describe it.

Reply:
Sounds good to me.